_posts/2019-12-06-totw-146.md

```diff
@@ -0,0 +1,222 @@
+---
+title: "Tip of the Week #146: Default vs Value Initialization"
+layout: tips
+sidenav: side-nav-tips.html
+published: true
+permalink: tips/146
+type: markdown
+order: "146"
+---
+
+Originally published as totw/146 on 2018-04-19
+
+*By Dominic Hamon [(dominic@google.com)](mailto:dominic@google.com)*
+
+Updated 2019-11-25
+
+Quicklink: [abseil.io/tips/146](https://abseil.io/tips/146)
+
+“The road to success is always under construction.” -- [Lily
+Tomlin](https://en.wikipedia.org/wiki/Lily_Tomlin)
+
+## TL;DR
+
+For safety and readability, you should assume scalar objects are not initialized
+to a reasonable value until they have been explicitly set to a value. Use of
+initializers can ensure that scalar values are initialized to safe values.
+
+## Introduction
+
+When objects are created, they may be initialized or uninitialized.
+Uninitialized objects are not safe to read, but understanding when the object is
+uninitialized is not trivial.
+
+The first thing to understand is if the type under construction is scalar,
+aggregate, or some other type. A *scalar* type can be thought of as a simple
+type: an integral or floating point arithmetic object; a pointer; an enum; a
+pointer-to-member; `nullptr_t`. An *aggregate* type is an array or trivial class
+(one with only public, non-static data members, no user-provided constructors,
+no base classes, no virtual member functions, and no default member
+initializers).
+
+The second factor in figuring out if an instance has been initialized to a value
```

Comment:
Awkward sentence:

"Another factor affecting whether an instance has been initialized to a value that is safe to read is whether it has an explicit initializer"?

Reply:
Done.